File: algorithm/LeetCode/CombinationSumIII/CombinationSumIII.cpp
class Solution {
public:
    vector<vector<int>> combinationSum3(int k, int n){
        size = k;
        target = n;
        vector<int> tmpVec;
        tmpVec.clear();
        ret.clear();
        dfs(tmpVec, target, 1);
        return ret;
    }
    void dfs(vector<int>& vec, int left, int start)
    {
        if(left < 0) return;
        if(left == 0 && vec.size() == size){ 
            ret.push_back(vec);
            return;
        }
        for(int i = start; i <= 9; ++i){
            vec.push_back(i);
            dfs(vec, left - i, i + 1);
            vec.pop_back();
        }
    }
private:
    vector<vector<int>> ret;
    int target;
    int size;
};

File: algorithm/LeetCode/CombinationSumIII/CombinationSumIII.java
public class Solution {
    public List<List<Integer>> combinationSum3(int k, int n) {
        List<List<Integer>> ret = new ArrayList<List<Integer>>();
        List<Integer> tmp = new ArrayList<Integer>();
        dfs(ret, tmp, 0, k, 1, n); 
        return ret;  
    }
    
    public void dfs(List<List<Integer>>ret, List<Integer>tmp, int dep, int maxDep, int start, int left){
        if(left < 0)
            return; //回溯
        else if(left == 0){
            if(dep == maxDep)
                ret.add(new ArrayList<Integer>(tmp));
            return;
        }else{
            for(int i = start; i <= 9; ++i){
                tmp.add(i);
                dfs(ret, tmp, dep+1, maxDep, i+1, left - i);
                tmp.remove(new Integer(i));
            }
        }
    }
}